### PR TITLE
HTTP/2 HPACK Header Name Validation and Trailing Padding

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
Motivation:
The HPACK code currently disallows empty header names. This is not explicitly forbidden by the HPACK RFC https://tools.ietf.org/html/rfc7541. However the HTTP/1.x RFC https://tools.ietf.org/html/rfc7230#section-3.2 and thus HTTP/2 both disallow empty header names, and so this precondition check should be moved from the HPACK code to the protocol level.
HPACK also requires that string literals which are huffman encoded must be treated as an encoding error if the string has more than 7 trailing padding bits https://tools.ietf.org/html/rfc7541#section-5.2, but this is currently not enforced.

Result:
- HPACK to allow empty header names
- HTTP/1.x and HTTP/2 header validation should not allow empty header names
- Enforce max of 7 trailing padding bits

Result:
Code is more compliant with the above mentioned RFCs
Fixes https://github.com/netty/netty/issues/5228

### CODE REVIEW

**Comment by carl-mastrangelo** [comment]
Merged from https://github.com/netty/netty/pull/5260 and https://github.com/golang/net/commit/6050c111928ef9186beb82e4395437607bb24bdb

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java [hunk 1]
```diff
@@ -214,17 +222,12 @@ public void decode(InputStream in, HeaderListener headerListener) throws IOExcep
                     } else {
                         nameLength = index;
 
-                        // Disallow empty names -- they cannot be represented in HTTP/1.x
-                        if (nameLength == 0) {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java]: why is this removed?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java [hunk 2]
```diff
@@ -502,9 +505,6 @@ private void insertHeader(HeaderListener headerListener, byte[] name, byte[] val
 
     private void addHeader(HeaderListener headerListener, byte[] name, byte[] value,
                            boolean sensitive) {
-        if (name.length == 0) {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java]: same question as above

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HpackUtil.java [hunk 3]
```diff
@@ -51,16 +48,6 @@ static boolean equals(byte[] s1, byte[] s2) {
         return c == 0;
     }
 
-    /**
-     * Checks that the specified object reference is not {@code null}.
-     */
-    static <T> T requireNonNull(T obj) {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HpackUtil.java]: Yes, thanks for cleaning this up :)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java [hunk 4]
```diff
@@ -31,13 +31,20 @@
...
 package io.netty.handler.codec.http2.internal.hpack;
 
+import io.netty.util.internal.EmptyArrays;
+
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 
 final class HuffmanDecoder {
 
-    private static final IOException EOS_DECODED = new IOException("EOS Decoded");
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: should you set the stack trace like you do in the other place?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java [hunk 5]
```diff
@@ -67,42 +74,50 @@
 
         Node node = root;
         int current = 0;
-        int bits = 0;
+        int currentBits = 0;
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: Can you add docs to these variables?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java [hunk 6]
```diff
@@ -67,42 +74,50 @@
...
+        int symbolBits = 0;
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: Add docs to this for loop?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java [hunk 7]
```diff
@@ -67,42 +74,50 @@
...
                     baos.write(node.symbol);
                     node = root;
+                    symbolBits = currentBits;
                 }
             }
         }
 
-        while (bits > 0) {
-            int c = (current << (8 - bits)) & 0xFF;
+        while (currentBits > 0) {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: Add docs to this for loop?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java (hunk 1 above)
Comment (carl-mastrangelo) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java]: Hpack allows zero length string headers, but HTTP does not.  The check moved to  DefaultHttp2Headers

**Comment by nmittler** [comment]
@carl-mastrangelo looks good! Can you just add some docs (see my comments)?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java (hunk 1 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java]: sgtm

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java (hunk 2 above)
Comment (carl-mastrangelo) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java]: Same reason, 0 is allowed for hpack, not for http.  (this is basically a technicality).

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HpackUtil.java (hunk 3 above)
Comment (carl-mastrangelo) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HpackUtil.java]: Thank Scott, I only fixed the hpack decoder!

File: codec-http/src/test/java/io/netty/handler/codec/http/DefaultHttpHeadersTest.java [hunk 8]
```diff
@@ -36,6 +36,16 @@
...
     private static final CharSequence HEADER_NAME = "testHeader";
 
+    @Test(expected = IllegalArgumentException.class)
+    public void nullHeaderNameNotAllowed() {
+        new DefaultHttpHeaders().add(null, "foo");
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void emtpyHeaderNameNotAllowed() {
+        new DefaultHttpHeaders().add("", "foo");
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/DefaultHttpHeadersTest.java]: StringUtil.EMPTY_STRING

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2HeadersTest.java [hunk 9]
```diff
@@ -30,6 +30,16 @@
...
 public class DefaultHttp2HeadersTest {
 
+    @Test(expected = Http2Exception.class)
+    public void nullHeaderNameNotAllowed() {
+        new DefaultHttp2Headers().add(null, "foo");
+    }
+
+    @Test(expected = Http2Exception.class)
+    public void emtpyHeaderNameNotAllowed() {
+        new DefaultHttp2Headers().add("", "foo");
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2HeadersTest.java]: StringUtil.EMPTY_STRING

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java (hunk 6 above)
Comment (carl-mastrangelo) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: Done.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java (hunk 5 above)
Comment (carl-mastrangelo) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: Done.

File: codec-http/src/test/java/io/netty/handler/codec/http/DefaultHttpHeadersTest.java (hunk 8 above)
Comment (carl-mastrangelo) [inline codec-http/src/test/java/io/netty/handler/codec/http/DefaultHttpHeadersTest.java]: Done.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2HeadersTest.java (hunk 9 above)
Comment (carl-mastrangelo) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2HeadersTest.java]: Done.

**Comment by carl-mastrangelo** [comment]
@nmittler @normanmaurer PTAL.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HpackUtil.java (hunk 3 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HpackUtil.java]: Thanks @Scottmitch :)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java [hunk 10]
```diff
@@ -65,44 +72,78 @@
...
         ByteArrayOutputStream baos = new ByteArrayOutputStream();
 
+        /*
+         * The idea here is to consume whole bytes at a time rather than individual bits.  node
+         * represents the Huffman tree, with all bit patterns denormalized as 256 children.  Each
+         * child represents the last 8 bits of the huffman code.  The parents of each child each
+         * represent the successive 8 bit chunks that lead up to the last most part.  8 bit bytes
+         * from buf are used to traverse these tree until a terminal node is found.
+         *
+         * current is a bit buffer.  The low order bits represent how much of the huffman code has \
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: nit: trailing "\" character

Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: Also remove double whitespace in front of `The`

File: codec-http2/src/test/java/io/netty/handler/codec/http2/internal/hpack/DecoderTest.java [hunk 11]
```diff
@@ -74,14 +81,63 @@ public void setUp() {
...
+            verify(mockListener, times(1)).addHeader(EMPTY_BYTES, EMPTY_BYTES, false);
+        } finally {
+            in.close();
+        }
+    }
+
+    @Test(expected = IOException.class)
+    public void testLiteralHuffmanEncodedWithPaddingGreaterThan7Throws() throws IOException {
+        byte[] input = {0, (byte) 0x81, -1};
+        ByteArrayInputStream in = new ByteArrayInputStream(input);
```
Comment (nmittler) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/internal/hpack/DecoderTest.java]: Could you separate out this into a separate utility method to clean up the test methods? It would be easier to read if we had something like:

```java
byte[] input = ...
decode(input);
verify(mockListener, times(1))...
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java [hunk 12]
```diff
@@ -65,44 +72,78 @@
...
+         * not been used to traverse the tree.  Thus, the high order bits are just garbage.
+         * currentBits represents how many of the low order bits of current are actually valid.
+         * currentBits will vary between 0 and 15.
+         *
+         * symbolBits is the number of bits of the the symbol being decoded, *including* all those
+         * of the parent nodes.  symbolBits tells how far down the tree we are.  For example, when
+         * decoding the invalid sequence {0xff, 0xff}, currentBits will be 0, but symbolBits will be
+         * 16.  This is used to know if buf ended early (before consuming a whole symbol) or if
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: Remove double whitespace in front of `This`-

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java [hunk 13]
```diff
@@ -65,44 +72,78 @@
...
+         * represent the successive 8 bit chunks that lead up to the last most part.  8 bit bytes
+         * from buf are used to traverse these tree until a terminal node is found.
+         *
+         * current is a bit buffer.  The low order bits represent how much of the huffman code has \
+         * not been used to traverse the tree.  Thus, the high order bits are just garbage.
+         * currentBits represents how many of the low order bits of current are actually valid.
+         * currentBits will vary between 0 and 15.
+         *
+         * symbolBits is the number of bits of the the symbol being decoded, *including* all those
+         * of the parent nodes.  symbolBits tells how far down the tree we are.  For example, when
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: remove double whitespace

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java [hunk 14]
```diff
@@ -65,44 +72,78 @@
     public byte[] decode(byte[] buf) throws IOException {
         ByteArrayOutputStream baos = new ByteArrayOutputStream();
 
+        /*
+         * The idea here is to consume whole bytes at a time rather than individual bits.  node
+         * represents the Huffman tree, with all bit patterns denormalized as 256 children.  Each
+         * child represents the last 8 bits of the huffman code.  The parents of each child each
+         * represent the successive 8 bit chunks that lead up to the last most part.  8 bit bytes
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: remove double whitespace

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java [hunk 15]
```diff
@@ -65,44 +72,78 @@
     public byte[] decode(byte[] buf) throws IOException {
         ByteArrayOutputStream baos = new ByteArrayOutputStream();
 
+        /*
+         * The idea here is to consume whole bytes at a time rather than individual bits.  node
+         * represents the Huffman tree, with all bit patterns denormalized as 256 children.  Each
+         * child represents the last 8 bits of the huffman code.  The parents of each child each
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: remove double whitespace

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java [hunk 16]
```diff
@@ -65,44 +72,78 @@
     public byte[] decode(byte[] buf) throws IOException {
         ByteArrayOutputStream baos = new ByteArrayOutputStream();
 
+        /*
+         * The idea here is to consume whole bytes at a time rather than individual bits.  node
+         * represents the Huffman tree, with all bit patterns denormalized as 256 children.  Each
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: remove double whitespace

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java [hunk 17]
```diff
@@ -65,44 +72,78 @@
     public byte[] decode(byte[] buf) throws IOException {
         ByteArrayOutputStream baos = new ByteArrayOutputStream();
 
+        /*
+         * The idea here is to consume whole bytes at a time rather than individual bits.  node
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: remove double whitespace

File: codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java [hunk 18]
```diff
@@ -65,44 +72,78 @@
...
+                    symbolBits = currentBits;
                 }
             }
         }
 
-        while (bits > 0) {
-            int c = (current << (8 - bits)) & 0xFF;
+        /*
+         * We have consumed all the bytes in buf, but haven't consumed all the symbols. We may be on
+         * a partial symbol, so consume until there is nothing left.  This will loop at most 2
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java]: remove double whitespace

**Comment by nmittler** [comment]
@carl-mastrangelo LGTM ... just a couple of nits.  Can you address and rebase?

**Comment by carl-mastrangelo** [comment]
@nmittler @normanmaurer PTAL. Also lots of Done.

**Comment by nmittler** [comment]
LGTM

**Comment by nmittler** [comment]
Cherry-picked as into 4.1 (1cb706a)

**Comment by carl-mastrangelo** [comment]
This can be closed now.

**Comment by Scottmitch** [comment]
Thanks @carl-mastrangelo @nmittler for taking care of the remaining issues!
